### PR TITLE
adding parrallel for loop

### PR DESCRIPTION
**Pull request by maorn**
i didn't change any logic, and tried to write it in the most clearer way 
let me know if there are any issues , i can change it again 
Maor

### CODE REVIEW

**Comment by dswah** [comment]
@maorn great!! 
i will try it on my machine today or tomorrow.

File: pygam/pygam.py [hunk 1]
```diff
@@ -87,22 +87,74 @@
...
                'monotonic_inc': monotonic_inc,
                'monotonic_dec': monotonic_dec,
                'circular': circular,
                'none': none
-              }
-
+               }
+
+
+def _parallel_inc(gam, X, y, weights):
```
Comment (dswah) [inline pygam/pygam.py]: why did you prefer to make this a function instead of a method of the GAM?

Comment (dswah) [reply pygam/pygam.py]: @maorn

File: pygam/pygam.py [hunk 2]
```diff
@@ -87,22 +87,74 @@
...
+    n = len(y)
+    R = []
+    f = []
+    r = 0.
+    vars_ = defaultdict(list)
+    pool = Pool()
+    block_results = []
+    for mask in gam._block_masks(n):
+        Xk = gam._modelmat(X[mask])
```
Comment (dswah) [inline pygam/pygam.py]: @maorn i suspect that we will need to move the line:
`Xk = gam._modelmat(X[mask])`
into the `process_one` function. this way we avoid forming.holding the entire model matrix in memory at any given time.

File: pygam/pygam.py [hunk 3]
```diff
@@ -87,22 +87,74 @@
...
+        block_results.append(pool.apply_async(process_one,
+                                              args=(gam, Xk, y[mask], weights[mask],)))
+    pool.close()
+    pool.join
+
+    res = [x.get() for x in block_results if x is not None]
+    res = list(filter(None.__ne__, res))
```
Comment (dswah) [inline pygam/pygam.py]: this line is causing failure in python 2.x because `None` doesn't have the `__ne__` method before python 3.
why do we need this line anyways? 
is it possible x is not None, but still x.get() returns None?

either way, perhaps we should just do another list comprehension?

```python
results = [x.get() for x in block_results if x is not None]
results = [result for result in results if result is not None]
```

Comment (dswah) [reply pygam/pygam.py]: @maorn

File: pygam/pygam.py [hunk 4]
```diff
@@ -87,22 +87,74 @@
...
+    Q, R = np.linalg.qr(np.vstack(R))
+    f = Q.T.dot(np.concatenate(f))
+    vars_['pseudo_data'] = np.concatenate(vars_['pseudo_data'])
+    vars_['mu'] = np.concatenate(vars_['mu'])
+    vars_['y'] = np.concatenate(vars_['y'])
+    return Q, R, f, r, vars_
+
+
+def process_one(gam, Xk, y_mask, weight_mask):
+    Xk, yk, zk, Wk, muk = gam._forward_pass(Xk, y_mask, weight_mask)
```
Comment (dswah) [inline pygam/pygam.py]: perhaps we pull in the 
```python
Xk = gam._modelmat(X[mask])
```
into here?

File: pygam/pygam.py [hunk 5]
```diff
@@ -1242,45 +1294,7 @@ def _incremental_pirls_parallel(self, X, y, weights):
...
-
-        # now combine all partitions
-        Q, R = np.linalg.qr(np.vstack(R))
-        f = Q.T.dot(np.concatenate(f))
-
-        vars_['pseudo_data'] = np.concatenate(vars_['pseudo_data'])
-        vars_['mu'] = np.concatenate(vars_['mu'])
-        vars_['y'] = np.concatenate(vars_['y'])
-        return Q, R, f, r, vars_
+        return _parallel_inc(self, X, y, weights)
```
Comment (dswah) [inline pygam/pygam.py]: @maorn same question here: why dont we just keep the logic in this method?

File: pygam/pygam.py [hunk 6]
```diff
@@ -2357,7 +2384,7 @@ def gridsearch(self, X, y, weights=None, return_scores=False,
         if progress:
             pbar = ProgressBar()
         else:
-            pbar = lambda x: x
+            def pbar(x): return x
```
Comment (dswah) [inline pygam/pygam.py]: @maorn 
whats the intuition behind this change?

File: pygam/pygam.py [hunk 7]
```diff
@@ -1305,23 +1319,24 @@ def _pirls(self, X, y, weights):
         # initialize GLM coefficients if model is not yet fitted
         if (not self._is_fitted or
             len(self.coef_) != sum(self._n_coeffs) or
-            not np.isfinite(self.coef_).all()):
+                not np.isfinite(self.coef_).all()):
 
-           # initialize the model
-           self.coef_ = self._initial_estimate(X, y)
+            # initialize the model
```
Comment (dswah) [inline pygam/pygam.py]: @maorn good catch.
yes, in general these formatting changes are helpful and appear to make the code more legible.

**Comment by dswah** [comment]
@maorn this is looking great.
i added a few questions.

Also i think it would be useful to add a `n_jobs=...` argument to the GAM constructor. And when `n_jobs==1` we should avoid running any parallelism. here is my reasonning:

-  i've read a few problems with [multiprocessing on OSX ('fork' vs 'spawn', ...)](http://scikit-learn.org/stable/faq.html#why-do-i-sometime-get-a-crash-freeze-with-n-jobs-1-under-osx-or-linux)
- we can use this to benchmark the speeds of parallel execution vs serial execution. in my own experiments i keep running into the typical case where parallelism doesnt help because [numpy already performs some parallelism on basic linear algebra operations.](https://stackoverflow.com/questions/15414027/multiprocessing-pool-makes-numpy-matrix-multiplication-slower)

also keeping a note here for later use
- [intel recommendations](https://software.intel.com/en-us/articles/large-matrix-operations-with-scipy-and-numpy-tips-and-best-practices)
- [dask backend](http://dask-ml.readthedocs.io/en/latest/hyper-parameter-search.html#flexible-backends)

File: pygam/pygam.py (hunk 1 above)
Comment (maorn) [reply pygam/pygam.py]: all objects that you pass in multiprocess (apply_async) should be  pickable (serialized, see here,
https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled) 
when you add multiprocess into a class there can be some issues with class members, so in general i always write the multiprocess part as an external function. 
you can insert it back and check that it works

File: pygam/pygam.py (hunk 2 above)
Comment (maorn) [reply pygam/pygam.py]: i am not sure how do you avoid it, since all the structures are in memory outside this code, there shouldn't be any structure(matrix)  coping so it won't explode the memory 
you can change the code here if needed

File: pygam/pygam.py (hunk 3 above)
Comment (maorn) [reply pygam/pygam.py]: these are general lines i use in all parallel methods for cases where you returned None in the function, in our case we always return a value.
so the second line can be removed

File: pygam/pygam.py (hunk 2 above)
Comment (dswah) [reply pygam/pygam.py]: @maorn ahh i think i was a little sleepy when i wrote this comment :)

every iteration in the loop is overwriting `Xk` so we are never keeping all the `Xk`s in memory at the same time. i guess it doesnt matter too much if this line goes inside or outside of the parallel loop... however, since `gam._modelmat()` has many non-linear algebra operations we indeed should see performance gains from parallelizing.

File: pygam/pygam.py (hunk 6 above)
Comment (dswah) [reply pygam/pygam.py]: ahh for pickling?

Comment (maorn) [reply pygam/pygam.py]: my eclipse autocorrects any pep8 and pylint errors and warnings. this is one of these changes (i didn't change it manually) 
basic explanation for this change (pep8 E731)
https://stackoverflow.com/questions/25010167/e731-do-not-assign-a-lambda-expression-use-a-def

File: pygam/pygam.py (hunk 1 above)
Comment (dswah) [reply pygam/pygam.py]: cool. i suspect that it should be OK as a method. 

since the argument `gam` would be the same as the argument `self` ?

**Comment by maorn** [comment]
Also i think it would be useful to add a n_jobs=... argument to the GAM constructor. And when n_jobs==1 we should avoid running any parallelism
Maor: this is a good idea , i used it my code, i didn't want to change it in your code ( i usually use 1.5*cpu_count() to keep my machine on 100% CPU). 
please notice that in case of n_jobs=1 you may want to run the function directly (avoid the Pool code) since running it with 1 will still activate the multi process with one process (hence no time will be saved).

File: pygam/pygam.py (hunk 1 above)
Comment (maorn) [reply pygam/pygam.py]: yes it should be fine, mind that all the classes that inherent from GAM class should be pickable as well to run this function

**Comment by dswah** [comment]
@maorn thanks very much for the code and the explanations. this is super helpful.

I have one last question:
what is the speedup required for us to use parallel code?

for example if this parallelism boosted speeds by 4x then we would probably like to use it.
but what if we only see a speed increase of 1%? 
then is it worth it? i think it becomes a more difficult to decide whether we want to maintain (improve, fix, debug, etc) additional functionality.

anyways... just something ive been thinkin about.

i still need to do a speed test.

did you do one?

**Comment by maorn** [comment]
i haven't done a speed test yet,do you run any profiling tools on this code? 
i would start by running profiler and then checking which code should be improved (or converted into parallel execution) 

i will try to run parallel test later this week

**Comment by dswah** [comment]
@maorn here are some small profiling experiments i tried
https://github.com/dswah/pyGAM/issues/190
